File: java/Arrays/Day10/ZeroOne.java
package Day10;
import java.util.*;

public class ZeroOne {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter the size of Array: ");
        int n = sc.nextInt();

        int arr[] = new int[n];
        System.out.print("Enter " + n + " " + "numbers: ");
        for (int i = 0; i < n; i++) {
            arr[i] = sc.nextInt();
        }



        // // better approach...
        // int cnt0 = 0; 
        // int cnt1 = 0; 
        // int cnt2 = 0;

        // for (int i = 0; i < n; i++) {
        //     if (arr[i] == 0) {
        //         cnt0++;
        //     } else if (arr[i] == 1) {
        //         cnt1++;
        //     } else {
        //         cnt2++;
        //     }
        // }

        // for (int i = 0; i < cnt0; i++) arr[i] = 0;
        // for (int i = cnt0; i < cnt0 + cnt1; i++) arr[i] = 1;
        // for (int i = cnt0 + cnt1; i < n; i++) arr[i] = 2;
        

        // System.out.print("Sorted array: ");
        // for (int num : arr) {
        //     System.out.print(num + " ");
        // }

        // // ( TC -> O(nlogn) and SC -> O(1) )





        // // Optimal approach :- 2 pointer (Dutch National Flag Algorithm)...
        int low = 0, mid = 0, high = n - 1;

        while (mid <= high) {
            if (arr[mid] == 0) {
                swap(arr, low, mid);
                low++; 
                mid++;
            } else if (arr[mid] == 1) {
                mid++;
            } else {
                swap(arr, mid, high);
                high--;
            }
        }


        System.out.print("Sorted Array with 2-pointer: ");
        for (int num : arr) {
            System.out.print(num + " ");
        }


        sc.close();
    }


    public static void swap(int[] arr, int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }
}

File: python/Arrays/day10/zeroOne.py
n = int(input("Enter the size of array: "))

arr = list(map(int, input(f"Enter {n} numbers: ").split()))




# # better approach...
# cnt0 = 0
# cnt1 = 0
# cnt2 = 0

# for i in range(n):
#     if arr[i] == 0:
#         cnt0 += 1
#     elif arr[i] == 1:
#         cnt1 += 1
#     else:
#         cnt2 += 1
    
# for i in range(cnt0):
#     arr[i] = 0

# for i in range(cnt0, cnt0 + cnt1):
#     arr[i] = 1

# for i in range(cnt0 + cnt1, n):
#     arr[i] = 2


# print("Sorted Array: ", *arr)





# # Optimal approach :- 2 pointer (Dutch National Flag Algorithm)...
low = 0
mid = 0
high = n - 1

while mid <= high:
    if arr[mid] == 0:
        arr[low], arr[mid] = arr[mid], arr[low]
        low += 1
        mid += 1
    elif arr[mid] == 1:
        mid += 1
    else:
        arr[mid], arr[high] = arr[high], arr[mid]
        high -= 1
    
print("Sorted Array with 2-pointer: ", *arr)